File: campaign_email_ReturnPath.r
# Created on Sun Oct 9 10:58:21 2016
# Author: Evelyn

#make sure the data file is in current work directoty
# prepare the workspace
rm(list = ls())
set.seed(2016)

library(ggplot2)
library(car)
library(caret)
library(randomForest)

# load data and look at data
data_orig <- read.csv(file = "assessment_challenge.csv")
dim(data_orig)
colnames(data_orig)
str(data_orig)
head(data_orig)
summary(data_orig)
#pairs(data_orig)

##########################################################
# Data pre-processing
##########################################################

#exempt records with missing values
data_nomissing <- na.omit(data_orig)

summary(data_nomissing$from_domain_hash)
summary(data_nomissing$Domain_extension)
summary(data_nomissing$day)

#exclude id from predictive analysis
data_no_id <- subset(data_nomissing, select = -id)

# from here we construct two data sets for separate modeling

#one is excluding Domain_extension and another excluding from_domain_hash
data_no_domain <- subset(data_no_id, select = -Domain_extension)

# include new levels into the level space of from_domain_hash
new_levels <- 0:20
levels(data_no_domain$from_domain_hash) <- c(levels(data_no_domain$from_domain_hash), new_levels)

#Combine levels of from_domain_hash according to response rate
# run on 5% sample data set
temp <- sample(nrow(data_no_domain), round(0.05*nrow(data_no_domain)))
sample_hash <- data_no_domain[temp,]
training <- data_no_domain[-temp,]
#calculate group mean of response rate by from_domain_hash
mean_by_domain <-aggregate(sample_hash$read_rate, 
                            by = list(sample_hash$from_domain_hash),FUN = mean, na.rm = TRUE)

for(i in 1:length(mean_by_domain$Group.1)){
  sample_hash$from_domain_hash[sample_hash$from_domain_hash == mean_by_domain$Group.1[i]] <- as.integer(round(mean_by_domain$x[i], 2) * 100 /5) + 1
}

# drop levels with 0 records
sample_hash$from_domain_hash <- factor(sample_hash$from_domain_hash)
sample_hash <- na.omit(sample_hash)

# modeling and prediction
# training set & testing set separation--0.8 for training, 0.2 for testing
train <- sample(nrow(sample_hash), round(0.8*nrow(sample_hash)))
training_sample_no_domain <- sample_hash[train,]
testing_sample_no_domain <- sample_hash[-train,]

#   1. Linear regression model
model_lr_sample<-lm(read_rate ~ ., data = training_sample_no_domain)

# check the model performance
print(summary(model_lr_sample))
opar <- par(mfrow = c(2,2), oma = c(0, 0, 1.1, 0))
plot(model_lr_sample, las = 1)

#training_rmse_lr_1<-sqrt(mean((model_lr_sample$residuals)^2))
predictions<-predict(model_lr_sample, testing_sample_no_domain)
testing_rmse_lr_1 <- sqrt(mean((testing_sample_no_domain$read_rate - predictions)^2))

#   2. Random Forest
model_rf_sample <- randomForest(read_rate ~ ., data = training_sample_no_domain, 
                         ntree = 100, mtry = 10,
                         importance= T, na.action = na.omit)
print(summary(model_rf_sample))
plot(model_rf_sample)
importance(model_rf_sample)
varImpPlot(model_rf_sample)

#training_rmse_rf_sample <- sqrt(mean((training_sample_no_domain$read_rate - model_rf_sample$predicted)^2))
predictions<-predict(model_rf_sample, testing_sample_no_domain)
testing_rmse_rf_sample <- sqrt(mean((testing_sample_no_domain$read_rate - predictions)^2))


#3. Random Forest, try to train only with most important predictors
model_rf_selected_sample <- randomForest(read_rate ~ from_domain_hash 
                                  + avg_user_domain_avg_read_rate + avg_domain_inbox_rate + avg_user_avg_read_rate, 
                                  data = training_sample_no_domain, 
                                  ntree = 100, mtry = 5,
                                  importance= T, na.action = na.omit)
print(summary(model_rf_selected_sample))
plot(model_rf_selected_sample)
importance(model_rf_selected_sample)
varImpPlot(model_rf_selected_sample)

#training_rmse_rf_selected <- sqrt(mean((training$day_on_market - model_rf_selected$predicted)^2))
predictions<-predict(model_rf_selected_sample, testing_sample_no_domain)
testing_rmse_rf_selected_sample <- sqrt(mean((testing_sample_no_domain$read_rate - predictions)^2))



#Another is excluding from_domain_hash from data set
data_no_hash <- subset(data_no_id, select = -from_domain_hash)
#remove anomalies in Domain_extension
data_no_hash_done<-subset(data_no_hash, !grepl("[[:digit:]]", Domain_extension))

# include new levels into the level space of Domain_extension
new_levels <- 0:20
levels(data_no_hash_done$Domain_extension) <- c(levels(data_no_hash_done$Domain_extension), new_levels)

#Combine levels of Domain_extension according to response rate
#calculate group mean of response rate by Domain_extension
mean_by_hash <-aggregate(data_no_hash_done$read_rate, 
                           by = list(data_no_hash_done$Domain_extension),FUN = mean, na.rm = TRUE)

for(i in 1:length(mean_by_hash$Group.1)){
  data_no_hash_done$Domain_extension[data_no_hash_done$Domain_extension == mean_by_hash$Group.1[i]] <- as.integer(round(mean_by_hash$x[i], 2) * 100 /5) + 1
}

# drop levels with 0 records
data_no_hash_done$Domain_extension <- factor(data_no_hash_done$Domain_extension)

# modeling and prediction
# training set & testing set separation--0.8 for training, 0.2 for testing
train <- sample(nrow(data_no_hash_done), round(0.8*nrow(data_no_hash_done)))
training_sample_no_hash <- data_no_hash_done[train,]
testing_sample_no_hash <- data_no_hash_done[-train,]

#   1. Linear regression model
model_lr_no_hash<-lm(read_rate ~ ., data = training_sample_no_hash)

# check the model performance
print(summary(model_lr_no_hash))
#opar <- par(mfrow = c(2,2), oma = c(0, 0, 1.1, 0))
#plot(model_lr_no_hash, las = 1)

#training_rmse_lr_1<-sqrt(mean((model_lr_sample$residuals)^2))
predictions<-predict(model_lr_no_hash, testing_sample_no_hash)
testing_rmse_lr_2 <- sqrt(mean((testing_sample_no_hash$read_rate - predictions)^2))

#   2. Random Forest
model_rf_no_hash <- randomForest(read_rate ~ ., data = training_sample_no_hash, 
                                ntree = 100, mtry = 10,
                                importance= T, na.action = na.omit)
print(summary(model_rf_no_hash))
plot(model_rf_no_hash)
importance(model_rf_no_hash)
varImpPlot(model_rf_no_hash)

#training_rmse_rf_sample <- sqrt(mean((training_sample_no_domain$read_rate - model_rf_sample$predicted)^2))
predictions<-predict(model_rf_no_hash, testing_sample_no_hash)
testing_rmse_rf_no_hash <- sqrt(mean((testing_sample_no_hash$read_rate - predictions)^2))


#3. Random Forest, try to train only with most important predictors
model_rf_selected_no_hash <- randomForest(read_rate ~ avg_user_domain_avg_read_rate + avg_domain_inbox_rate + avg_domain_read_rate + mb_superuser + avg_user_avg_read_rate, 
                                         data = training_sample_no_hash, 
                                         ntree = 100, mtry = 5,
                                         importance= T, na.action = na.omit)
print(summary(model_rf_selected_no_hash))
plot(model_rf_selected_no_hash)
importance(model_rf_selected_no_hash)
varImpPlot(model_rf_selected_no_hash)

predictions<-predict(model_rf_selected_no_hash, testing_sample_no_hash)
testing_rmse_rf_selected_sample <- sqrt(mean((testing_sample_no_hash$read_rate - predictions)^2))

File: campaign_email_ReturnPath.py
# -*- coding: utf-8 -*-
"""
Created on Sat Oct 08 12:05:19 2016

@author: Evelyn
"""

import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
matplotlib.style.use('ggplot')
from sklearn.cross_validation import train_test_split
from sklearn import preprocessing
from sklearn.ensemble import RandomForestRegressor
from sklearn import linear_model

# load data
data = pd.read_csv('assessment_challenge.csv')
# check amounts of variables and records
print data.shape
# check missing values
print np.sum(data.isnull())

'''
*******************************************************************************
1--Cleaning, exploration, visualization and preparation of data
*******************************************************************************
'''
#exempt records with missing values
data_nomissing = data.dropna()
# check dimension again
print data_nomissing.shape

#check types of all columns
print data_nomissing.dtypes

#then check levels of all categorical variables
print len(data_nomissing.from_domain_hash.unique())
print len(data_nomissing.Domain_extension.unique())
print len(data_nomissing.day.unique())

#exclude id from predictive analysis
data_no_id = data_nomissing.drop(['id'], 1)

# from here we construct two data sets for separate modeling
#one is excluding Domain_extension and another excluding from_domain_hash
data_no_domain = data_no_id.drop(["Domain_extension"], 1)


#Combine levels of from_domain_hash according to response rate
#calculate group mean of response rate by from_domain_hash
mean_by_domain = data_no_domain["read_rate"].groupby(data_no_domain["from_domain_hash"]).mean()
print len(mean_by_domain.unique())

#construct a dictionary for lookup
domain_rate = {}
for domain in mean_by_domain.index:
    domain_rate[domain] = mean_by_domain[domain]

#data_no_domain["from_domain_hash"] = int((domain_rate[data_no_domain["from_domain_hash"]].round(2))*100)/5 + 1
#data_no_domain["from_domain_hash"].apply(int((domain_rate[data_no_domain["from_domain_hash"]].round(2))*100)/5 + 1)

for domain in mean_by_domain.index:
    data_no_domain.from_domain_hash[data_no_domain["from_domain_hash"] == domain] = int((mean_by_domain[domain].round(2))*100)/5 + 1

print data_no_domain.head()
print data_no_domain.shape
print len(data_no_domain.from_domain_hash.unique())

data_no_domain['from_domain_hash'] = data_no_domain['from_domain_hash'].astype('category')
print data_no_domain.dtypes

#encode categorical variables of data_no_domain and day
domain = preprocessing.LabelEncoder()
domain.fit(data_no_domain['from_domain_hash'])
data_no_domain['from_domain_hash'] = domain.transform(data_no_domain['from_domain_hash'])

day = preprocessing.LabelEncoder()
day.fit(data_no_domain['day'])
data_no_domain['day'] = day.transform(data_no_domain['day'])

#Linear regression-data_no_domain
X_train_s, X_test_s, y_train_s, y_test_s = train_test_split(data_no_domain.ix[:, 1:], 
                                                            data_no_domain['read_rate'], test_size=0.2, random_state=0)
lm_no_domain = linear_model.LinearRegression()
lm_no_domain.fit(X_train_s, y_train_s)

# The coefficients
print('Coefficients: \n', lm_no_domain.coef_)
# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((lm_no_domain.predict(X_test_s) - y_test_s) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((lm_no_domain.predict(X_test_s) - y_test_s) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % lm_no_domain.score(X_test_s, y_test_s))

#Random Forest for data_no_domain
model_RF_s = RandomForestRegressor(n_estimators = 1000)
model_RF_s.fit(X_train_s, y_train_s)
#results = model_RF_s.predict(X_test_s)

# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((model_RF_s.predict(X_test_s) - y_test_s) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((model_RF_s.predict(X_test_s) - y_test_s) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % model_RF_s.score(X_test_s, y_test_s))

importances = model_RF_s.feature_importances_
feature_importance = 100.0 * (importances / importances.max())
sorted_idx = np.argsort(feature_importance)
pos = np.arange(sorted_idx.shape[0]) + .5
plt.barh(pos, feature_importance[sorted_idx], align = 'center')
arr = np.array(list(data_no_domain.columns.values)[1:])
plt.yticks(pos, arr[sorted_idx])
plt.xlabel('Relative Importance')
plt.title('Feature Importance')
plt.show()

#Random Forest for data_no_domain only use important factors
X_test_new_s = data_no_domain[['from_domain_hash', 'avg_domain_inbox_rate', 'avg_user_avg_read_rate', 'avg_user_domain_avg_read_rate']]
X_train_new_s, X_test_new_s, y_train_new_s, y_test_new_s = train_test_split(X_test_new_s, data_no_domain['read_rate'], 
                                                    test_size = 0.2, random_state = 100)
model_RF_new_s = RandomForestRegressor(n_estimators = 1000)
model_RF_new_s.fit(X_train_new_s, y_train_new_s)

# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((model_RF_new_s.predict(X_test_new_s) - y_test_new_s) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((model_RF_new_s.predict(X_test_new_s) - y_test_new_s) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % model_RF_new_s.score(X_test_new_s, y_test_new_s))

#another way; only keep Domain_extension
#remove anomalies in Domain_extension
data_no_digit = data_no_id[data_no_id.Domain_extension.str.isdigit() == False]
data_no_digit = data_no_digit[data_no_digit.Domain_extension.str.startswith("0") == False]
data_no_digit = data_no_digit[data_no_digit.Domain_extension.str.startswith("?") == False]
data_no_hash = data_no_digit.drop(["from_domain_hash"], 1)
print len(data_no_hash.Domain_extension.unique())
print data_no_hash.head()

#merge levels of Domain_extension based on response rate
mean_by_domain_name = data_no_hash["read_rate"].groupby(data_no_hash["Domain_extension"]).mean()
print len(mean_by_domain_name.unique())

for domain in mean_by_domain_name.index:
    data_no_hash.Domain_extension[data_no_hash["Domain_extension"] == domain] = int((mean_by_domain_name[domain].round(2))*100)/5 + 1

print data_no_hash.head()
print len(data_no_hash.Domain_extension.unique())
data_no_hash['Domain_extension'] = data_no_hash['Domain_extension'].astype('category')
print data_no_hash.dtypes

#encode categorical variables of data_no_hash
domain = preprocessing.LabelEncoder()
domain.fit(data_no_hash['Domain_extension'])
data_no_hash['Domain_extension'] = domain.transform(data_no_hash['Domain_extension'])

day = preprocessing.LabelEncoder()
day.fit(data_no_hash['day'])
data_no_hash['day'] = day.transform(data_no_hash['day'])

print data_no_hash.dtypes

#Linear regression-data_no_hash
X_train_t, X_test_t, y_train_t, y_test_t = train_test_split(data_no_hash.ix[:, 1:], data_no_hash['read_rate'], test_size=0.2, random_state=0)
lm_no_hash = linear_model.LinearRegression()
lm_no_hash.fit(X_train_t, y_train_t)
# The coefficients
print('Coefficients: \n', lm_no_hash.coef_)
# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((lm_no_hash.predict(X_test_t) - y_test_t) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((lm_no_hash.predict(X_test_t) - y_test_t) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % lm_no_hash.score(X_test_t, y_test_t))

#Random Forest for data_no_hash
model_RF_t= RandomForestRegressor(n_estimators = 1000)
model_RF_t.fit(X_train_t, y_train_t)

# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((model_RF_t.predict(X_test_t) - y_test_t) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((model_RF_t.predict(X_test_t) - y_test_t) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % model_RF_t.score(X_test_t, y_test_t))

importances = model_RF_t.feature_importances_
feature_importance = 100.0 * (importances / importances.max())
sorted_idx = np.argsort(feature_importance)
pos = np.arange(sorted_idx.shape[0]) + .5
plt.barh(pos, feature_importance[sorted_idx], align = 'center')
arr = np.array(list(data_no_hash.columns.values)[1:])
plt.yticks(pos, arr[sorted_idx])
plt.xlabel('Relative Importance')
plt.title('Feature Importance')
plt.show()

#Random Forest for data_no_hash only use important factors
X_test_new_t = data_no_hash[['avg_domain_read_rate', 'avg_domain_inbox_rate', 'avg_user_avg_read_rate', 'avg_user_domain_avg_read_rate']]
X_train_new_t, X_test_new_t, y_train_new_t, y_test_new_t = train_test_split(X_test_new_t, data_no_hash['read_rate'], 
                                                    test_size = 0.2, random_state = 100)
model_RF_new_t = RandomForestRegressor(n_estimators = 1000)
model_RF_new_t.fit(X_train_new_t, y_train_new_t)

# The mean squared error
print("Mean squared error: %.4f"
      % np.mean((model_RF_new_t.predict(X_test_new_t) - y_test_new_t) ** 2))

print("Root mean squared error: %.4f"
      % (np.mean((model_RF_new_t.predict(X_test_new_t) - y_test_new_t) ** 2))**0.5)

# Explained variance score: 1 is perfect prediction
print('Variance score: %.4f' % model_RF_new_t.score(X_test_new_t, y_test_new_t))



